maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/MavenWorkspaceReader.java

```diff
@@ -28,4 +29,20 @@
 public interface MavenWorkspaceReader extends WorkspaceReader {
 
     Model findModel(Artifact artifact);
+
+    /**
+     * List all available artifacts this workspace repository manages.
+     *
+     * @return a stream of artifacts in no particular order
+     * @since 3.9.0
+     */
+    Stream<Artifact> listArtifacts();
```

Comment:
maybe it requires a list (both methods) cause in current state nobody closes the related streams so it can lead to issues (rest LGTM)

Reply:
[From the javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html):

> Streams have a [BaseStream.close()](https://docs.oracle.com/javase/8/docs/api/java/util/stream/BaseStream.html#close--) method and implement [AutoCloseable](https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html), but nearly all stream instances do not actually need to be closed after use. Generally, only streams whose source is an IO channel (such as those returned by [Files.lines(Path, Charset)](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#lines-java.nio.file.Path-java.nio.charset.Charset-)) will require closing. Most streams are backed by collections, arrays, or generating functions, which require no special resource management.

as there is no IO involved, there is no problem and callers can convert to list if they think it is required.

Reply:
Why the not return a collection? A conversion to a stream is still possible.

Reply:
Well that's all a bit hypothetical.... if you are certain you can always wrap streams in try-with-resource... 
Streams are much better suited than collections, as the caller can filter them or convert them to any type they need without having to think about if the collection could be modified or not.

Streams also the the advantage that they are _lazy_ if I search for an item (using `findFirst()` / `findAny()` / `anyMatch()`) only those items are evaluated up to the point where a match is found in contrast to a collection that is _greedy_ and thus requires to evaluate all content even if the first item is what I want.

Reply:
Streams also have the con to need collection to be cached (if you call it twice for ex).

Streams also have the disavantage to not be reusable so if anyone call it and need to browse it twice it will have to materialize it too (toList()) making the call.

So at the end from a caller perspective the collection usage is lighter in terms of contract and compatible with streams (`.stream()`).

now if you explicit you must enforce close - which is the current PR contract since everything is abstracted and can be backed by a filesystem by contract - it is fine too, just not as convenient IMHO.

Key decision point for me is: does the data fit in mem: obviously yes it is in maps already, so no need to stream it IMHO.

Reply:
Caching is not a concern of the API. Using streams allows implements to cache (as they can be sure no one can ever modify the underlying collection, what ever type and whatever technique). so this is really not an argument.

If I need to browse them twice (why?) I can again make a collection, I even need to do so, because I can not assume it is always stable (e.g. modified) unless I copy it (or the calling method copies it) so this is really not an advantage at all.

> So at the end from a caller perspective the collection usage is lighter in terms of contract and compatible with streams (.stream()).

I don't think so, and of course stream is compatible with collection as well (.toList()) so this also is just not really an argument.

> now if you explicit you must enforce close 

Sorry the statement of that one **must** close a stream was brought up by you and I proved its wrong, the API clearly state that closing is not required except for very special (and documented) cases, and even there it is documented that one _should_ close the stream see: https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#lines-java.nio.file.Path-java.nio.charset.Charset-

> If timely disposal of file system resources is required, the try-with-resources construct should be used to ensure that the stream's [close](https://docs.oracle.com/javase/8/docs/api/java/util/stream/BaseStream.html#close--) method is invoked after the stream operations are completed.

> now if you explicit you must enforce close - which is the current PR contract since everything is abstracted and can be backed by a filesystem by contract - it is fine too, just not as convenient IMHO.

so this whole "close" story is completely unrelated here and is just confusing, there are literally no code except the I/O cases where it is recommended, and you won't find code that closes simple streams "just in case"... and even if an implementation might has requirements for close it still can collect everything, close the stream and return a "save" stream to the caller, but that's nothing an API must take care of.

> Key decision point for me is: does the data fit in mem: obviously yes it is in maps already, so no need to stream it IMHO.

This is obviously not a decisions point, because then most of the time no one would ever use streams as most of them "fit in memory". And even if the current one is held in memory this won't necessary be the case for other implementations of this and it might be costly to collect everything, it even might delay thing (just think about maven would resolve and download **all** artifacts before the progress is reported in one big bunch, even though it is same as fast, you will want a stream of messages as they happen)

Also please look at the  `MavenChainedWorkspaceReader` implementation, that currently can simply return the stream from inner providers, combine them, and even can make sure they are still unique, while a collection approach would require:

- Each implementation must collect everything eager (even if not used at all)
- It must create a copy (if dynamic) or read only collection
- The chained reader must again maintain own collections and add all element it collects again, making it unique e.g. by a set (additional efforts) that includes traversing all object
- The caller most likely will not just store the collection but iterate it (foreach or stream) and thus must traverse all objects again.

Because of this, I'm strongly convinced that Streams are superior in this case (and even others), are a more modern way of "collections" and are much more flexible anyways, for everyone that is more used to collections they can easily be transformed into such.

Reply:
Very nice and profound knowledge you both have!

Reply:
> Caching is not a concern of the API.

iterating and reitarating over the map on big project will be a topic IMHO if these methods are useful.

> If I need to browse them twice (why?) I can again make a collection,

`I` is you in this PR? if so I agree, if the caller then no cause you can get multiple callers and would mean your API is not relevant.

> there are literally no code except the I/O cases where it is recommended

huh, no. But anyway you can't assume the related API has no IO related.

> This is obviously not a decisions point, because then most of the time no one would ever use streams as most of them "fit in memory".

Which is what is in most real  not overkill/hype API ;).

> while a collection approach would require

this part is wrong, what you propose is to visit the whole stream and mutate it whereas you can push back the filters and things like that in the actual materialization stream (the impl) which is what is done at maven in general (think scope filter for ex) and is more efficient on all the points you mention.

So ultimately, except the point we didn't understand each others, everything converges to a collection it seems.

Reply:
> So ultimately, except the point we didn't understand each others, everything converges to a collection it seems.

It does not, implementors are free to cache items internally in what ever way they like, this must not be exposed as API, callers are also free to hold their own copy in whatever type they seems fit their needs, this must not exposed in the API. Collection return types on the other hand has major drawbacks that need to be documented and mitigated streams do not have that are:

1. Carefully document if the returned collection is the property of the caller or the callee
2. Carefully document if the returned collection is safe to be shared across threads
3. Carefully document if the returned collection is safe to be cached
4. Carefully document if it is modifiable or not (just see recent bug in 3.8.6 ([MNG-7316](https://issues.apache.org/jira/browse/MNG-7316))

And yes I proposed the API and I designed the API and no one ever cared to do so before so it was designed with what I have in mind here and like to use, collections are not, especially not just for the sake "because we always did it that way" or some strange assumtions, otherwhise all Maven methods need to throw `Exception` and return AutoClosable objects because you "never can know". Thats why we have API contracts and the Contract says that a Stream (with no particular order) must be returned, it does not claim that the Stream must be closed nor that it throws `IOException` so if *any* I/O is performed as part of the implementation it is the fully responsibility of this hypothetical implementation to take action that ensure appropriate closing of system resources is performed and I/O problems are properly handled.

Reply:
> It does not, implementors are free to cache items internally

of reactor reader? you want to override it?

About the collection pitfalls, I guess it is mainly about being consistent with the track on the API and internals we started so I'd say:

1. make it immutable and in the javadoc,
2. solved by 1,
3. solved by 1,
4. solved by 1.

Strictly speaking streams have the exact same pitfall but you move this issue "when you materialize" the view so this is not really a point except the fact the view should be immutable as all the work done in the rest of the API probably.